src/lib/paths.ts

```diff
@@ -65,3 +71,51 @@ export async function getProjectMetaByRouteId(
 
   return route
 }
+
+const parseProjectRoute = (
+  configuration: Partial<SaveSettingsPayload>,
+  id: string
+) => {
+  const onDesktop = isDesktop()
+  let projectName = ''
+  let projectPath = ''
+  let currentFileName = ''
+  if (
+    configuration.app?.projectDirectory &&
+    id.startsWith(configuration.app.projectDirectory)
+  ) {
+    const relativeToRoot = window.electron.path.relative(
+      configuration.app.projectDirectory,
+      id
+    )
+    projectName = relativeToRoot.split(window.electron.path.sep)[0]
+    projectPath = window.electron.path.join(
+      configuration.app.projectDirectory,
+      projectName
+    )
+  } else {
+    projectPath = id
+    if (onDesktop) {
+      if (window.electron.path.extname(id) === '.kcl') {
+        projectPath = window.electron.path.dirname(id)
+      }
+      projectName = window.electron.path.basename(projectPath)
+    } else {
+      if (id.endsWith('.kcl')) {
+        projectPath = '/browser'
+        projectName = 'browser'
+      }
+    }
+  }
+  if (onDesktop) {
+    currentFileName = window.electron.path.basename(id)
+  } else {
+    currentFileName = 'main.kcl'
+  }
+  return {
+    projectName: projectName,
+    projectPath: projectPath,
+    currentFileName: currentFileName,
+    currentFilePath: id,
+  }
+}
```

Comment:
can we unit test in the same way as the rust tests were, only becaiuse there were lots of regression tests in those unit tests, aka those little shits are load bearing :)

Reply:
Yeah, I can unit test as-is, I took a lot of stuff out because it seems unnecessary. Like, the old code was _not_ using half of the function because of a bug with how config was passed in a json, we were just getting lucky before that this worked even with unit tests.

Reply:
If you know any of the specific functionality regressions I could write e2e tests for those too, to make sure I didn't cut out currently important stuff.

Reply:
mostly files w spaces, either url encoded or not were weird

Reply:
idk if applies here